### ISSUE TITLE
Type checking Uint32Array

### ISSUE THREAD

**Issue by saghul** [svaarala/duktape#752]
Hi!

I'm passing a Uint32Array down to C code and I wanted to do some type-checking, and check that the passed array is of that type, but all checks I can come up with by looking at the API don't seem to work. Is there something I'm missing?

Thanks!

**Comment by svaarala** [comment]
Are you looking for the C equivalent for the Ecmascript code?

```js
arg instanceof Uint32Array
```

If so, the C equivalent would be:

```c
duk_get_global_string(ctx, "Uint32Array");
if (!duk_instanceof(ctx, idx_arg, -1)) { ... }
duk_pop(ctx);
```

**Comment by svaarala** [comment]
It might be good if there was an API to get the bufferobject type (matching `duk_push_buffer_object()` typing).

**Comment by saghul** [comment]
> duk_get_global_string(ctx, "Uint32Array");
> if (!duk_instanceof(ctx, idx_arg, -1)) { ... } 
> duk_pop(ctx);

This is a bit too involved, I just wanted to place an assert, since it's an internal function.

> It might be good if there was an API to get the bufferobject type (matching duk_push_buffer_object() typing).

Yep, this is what I was after :-)

**Comment by svaarala** [comment]
A few notes from earlier API design considerations:

- Instead of providing a "get buffer object type" perhaps expose a `duk_get_class()` which would return a class number. The class number identifies buffer object types and also distinguishes between e.g. objects, arrays, RegExps, and functions: https://github.com/svaarala/duktape/blob/master/src/duk_hobject.h#L77-L107. The downside is that not all interesting "instance" relations are covered by the class number and that it's somewhat of an internal thing.
- Providing specific type testers like `duk_is_uint32array()` would make call sites more readable (at no footprint cost because they'd be macros) and would be consistent with how error types are handled (e.g. `duk_is_syntax_error()`).

**Comment by saghul** [comment]
FWIW, I like the approach. The second point though (adding `duk_is_uint32array` and so on) not so much. It would add a lot of noise to the API...

I was thinking on something like:

~~~~
duk_is_class(ctx, 0, DUK_TYPE_UINT32ARRAY)
~~~~

Or something along those lines.

**Comment by svaarala** [comment]
Would `duk_is_class(ctx, 0, DUK_TYPE_UINT32ARRAY)` be much better than the following?

```c
duk_get_class(ctx, 0) == DUK_TYPE_UINT32ARRAY
```

**Comment by saghul** [comment]
Nope, they would be the same :-) In my mind the is check is more aligned with other API functions, but that's up to you of course.

**Comment by svaarala** [comment]
Two similar calls are `duk_check_type()` and `duk_check_type_mask()`. The former allows checking for a single type and the latter for multiple types by specifying a bit mask.

**Comment by saghul** [comment]
Right. Maybe those could be extended then?

**Comment by svaarala** [comment]
A few more related/similar calls:

- duk_get_type() and duk_get_type_mask(): for primitive types, based on type tag
- duk_check_type() and duk_check_type_mask(): for checking primitive type(s), based on type tag mask
- duk_get_error_code(): get error code; this is based on error subclass relations (rather than a type tag or a class number)
- duk_is_syntax_error(), duk_is_range_error(), etc: error code checkers; no mask support for these

There are also some checks which are based on purely internal flags. For example, `duk_is_ecmascript_function()` checks the argument is a function and then checks the internal flag indicating the function is bytecode-based rather than a Duktape/C function.

**Comment by svaarala** [comment]
@saghul The difficulty is that there are many different "type" concepts which are not trivial to merge:

- Primitive types (`DUK_TYPE_STRING` etc) are type tags for plain types. This is the top level typing.
- Object classes are relevant only for objects (`DUK_TYPE_OBJECT`) and distinguish between arrays, plain objects, regexps, buffer object types, etc.
- However, object class does not always fully identify a type (for example `duk_is_ecmascript_function()`); also the case for user defined types.

From time to time I've tried to figure out how to simplify this but it seems difficult to have a single type concept; the top level type tag is needed to deal with the value stack correctly while in other cases one just needs to know if a type is "compatible" with some operation. For example, `duk_get_buffer_data()` relies on the target being any type with a relevant buffer (plain buffer or a buffer object).

Perhaps it'd be possible to achieve a two-level type structure where tag type is the top level type identifier and the second level would be a superset of object classes, error subclass identification, etc.

**Comment by svaarala** [comment]
By the way, one example of where type checking gets a bit hairy: what should `duk_is_uint32array()` return for an object inheriting from a Uint32Array instance?

From an operations/methods point of view it would behave like a Uint32Array in Ecmascript code. In C code, `duk_get_buffer_data()` wouldn't currently return a data pointer because the immediate object argument is not a typed array. Should a "is Uint32Array" primitive identify the argument based on an inheritance relationship or the type of the object itself?

The API is probably not very consistent in this respect right now. It's a good time to discuss this though because some changes can be made in 2.x.

**Comment by fatcerberus** [comment]
Hm, inheritance is a bit of a tricky thing.  Inheritance in JavaScript is prototypal so that you can have an "inherited" object which shares nothing in common with its base class other than a prototype identity.  Maybe they don't even have a constructor in common.  So even if `obj instanceof Uint32Array`  there might not even be an underlying buffer for Duktape to return!

I think Duktape rejecting inherited buffer objects is probably the most sane thing it can do.
